<img src="https://raw.githubusercontent.com/gustavogarzonuis/prueba/master/banner_IA.png" width="1000px" height="200px">

# QUIZ 02 - INTELIGENCIA ARTIFICIAL

## <span style="color:red"> Outline </span> 

1. [Ejercicio 1. Clasificación](#eje1)
2. [Ejercicio 2. Regresion](#eje2)
3. [Ejercicio 3. class Fashion](#eje3)
4. [Ejercicio 4. DNN](#eje4)

#### <font style="color: red"> Esta práctica contiene 4 ejercicios <!-- ONLY IN NOTEBOOK -->

<h2><span style="color:#FFF; background-color:#F00; font-weight:bold; padding:4px;">¡IMPORTANTE!</span> Ejecute las siguientes celdas para iniciar la actividad</h2>

In [1]:
import os
from google.colab import drive
drive.mount('/gdrive')
%cd /gdrive/My Drive/ai-q02
print(os.getcwd())
!ls

Drive already mounted at /gdrive; to attempt to forcibly remount, call drive.mount("/gdrive", force_remount=True).
/gdrive/My Drive/ai-q02
/gdrive/My Drive/ai-q02
convert_ipynb.sh  data	quiz_02.ipynb  README.md  run  utils


In [2]:
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
import pandas as pd

In [3]:
!pwd && chmod 777 convert_ipynb.sh run utils/*

/gdrive/My Drive/ai-q02


<h2><span style="color:#FFF; background-color:#F00; font-weight:bold; padding:4px;">¡IMPORTANTE!</span></h2><br><h2 style="margin-top:-10px;">Usted debe ejecutar <span style="color:#F00;">TODAS</span> las celdas de tipo <span style="color:#F00;">CODE</span>. De lo contrario, sus respuestas no serán procesadas correctamente.</h2>

---

### Ejercicio 1  <a name="eje1"></a>

En machine learning un factor importante es utilizar conjunto de datos clásicos y evaluar el comportamiento de los clasificadores. Teniendo en cuenta lo anterior, construya una función que: 

1. Cargue el dataset iris(), disponible en la libreria  `sklearn`
2. Cree dos clasificadores siguiendo la siguiente configuración: 
    1. Una máquina de soporte vectorial (SVC) con un kernel polinomial (poly) de grado 7, y `gamma='auto'`
    2. Un RandomForestClassifier con 100 estimadores y una profundidad máxima de 10 niveles
4. Clasifique el dataset con los dos clasificadores configurados.
    - Utilice una única partición de los datos, tomando el 35% como test
5. La función debe retornar los objetos SVM y RF entrenados. También una lista de los score obtenidos de la forma: `[score_rf, score_sv]`

Referencia SVC: [ https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html ] <br>
Referencia RandomForestClassifier: [ https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html ]

In [4]:
def fun_q1():
    from sklearn.ensemble import RandomForestClassifier
    from sklearn.svm import SVC
    from sklearn.model_selection import train_test_split
    from sklearn.datasets import load_iris
    import numpy as np
    
    iris = load_iris()
    X = iris.data
    y = iris.target

    X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=.35)
    a = SVC(kernel='poly', degree=7, gamma='auto')
    b=RandomForestClassifier(n_estimators=100, max_depth=10)
    a.fit(X_train,y_train)
    b.fit(X_train,y_train)
    c=a.score(X_test,y_test)
    d=b.score(X_test,y_test)
    return a, b, [c, d]

#import urllib, inspect
#src1 = urllib.quote_plus(inspect.getsource(fun_q1))
import urllib.parse, inspect
src1 = urllib.parse.quote_plus(inspect.getsource(fun_q1))

Pruebe su código:

In [5]:
svm, rf, scores = fun_q1()
print(svm.kernel,svm.degree, type(svm.kernel), type(svm.degree))
print(rf.max_depth)
print(scores)

poly 7 <class 'str'> <class 'int'>
10
[0.9811320754716981, 0.9811320754716981]


#### <span style="color:#FFF; background-color:#080; font-weight:bold; padding:4px;">Importante</span> Evalue su respuesta.

In [6]:
!./run CHECK_SOLUTION PQ2_5 $src1

evaluation result CORRECT


---

### Ejercicio 2  <a name="eje2"></a>

Completa la siguiente función para que realice una regresión utilizando un  **DecisionTreeRegressor** pasando como argumento **X_train** &nbsp; e &nbsp; **y_train**.<br>La función debe devolver el **objeto del regresor** y el **resultado** usando la siguiente configuración: 


- El regresor `DecisionTreeRegressor` debe ser entrenado con los parametros de `max_depth=8` y `criterion='mse'`
- La evaluación del método debe ser exhaustiva, utilizando un esquema de *cross-validation* con un **KFold = 20**
- La metrica para la evaluación es el MSE. Recuerde que en sklearn, se cuenta con la métrica MSE. 
- Se debe retornar el valor promedio de la métrica MSE. 
- se debe retornar el estimador ya entrenado

Referencia DecisionTreeRegressor [ https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeRegressor.html ]

In [7]:
def fit_reg(X_train, y_train):
    import numpy as np
    import pandas as pd
    from sklearn.tree import DecisionTreeRegressor
    from sklearn.model_selection import cross_val_score
    from sklearn.model_selection import KFold
    from sklearn.metrics import make_scorer, mean_squared_error

    est=DecisionTreeRegressor(max_depth=8, criterion='mse')
    s = cross_val_score(est, X_train, y_train, cv=KFold(20, shuffle=True), scoring=make_scorer(mean_squared_error))
    score=np.mean(s)
    return est, score

#import urllib, inspect
#src6 = urllib.quote_plus(inspect.getsource(fit_reg))
import urllib.parse, inspect
src6 = urllib.parse.quote_plus(inspect.getsource(fit_reg))

Cargue el dataset de bikes y pruebe la función realizada para estimar `count` (número de bicicletas) y usando como vector de entrenamiento, las caracteristicas: `temperature, humidity, windspeed`

In [8]:
import pandas as pd
bikes = pd.read_csv('data/bikes.csv')
bikes.head()

date  temperature   humidity  windspeed  count
0  2011-01-03     2.716070  45.715346  21.414957  120.0
1  2011-01-04     2.896673  54.267219  15.136882  108.0
2  2011-01-05     4.235654  45.697702  17.034578   82.0
3  2011-01-06     3.112643  50.237349  10.091568   88.0
4  2011-01-07     2.723918  49.144928  15.738204  148.0

In [9]:
import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeRegressor

est, s = fit_reg(bikes.values[:,1:4], bikes.values[:,-1])
print("est", est)
print("RMSE score", s)

est DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=8,
                      max_features=None, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, presort='deprecated',
                      random_state=None, splitter='best')
RMSE score 68239.2812883466


#### <span style="color:#FFF; background-color:#080; font-weight:bold; padding:4px;">Importante</span> Evalue su respuesta.

In [10]:
!./run CHECK_SOLUTION PQ2_6 $src6

71381.65045780611 68112.32932873405
evaluation result CORRECT


---

### Ejercicio 3  <a name="eje3"></a>

Vamos ahora a evaluar los algoritmos vistos en clase con el dataset **Fashion-MNIST**. Este dataset tiene un conjunto de clases 

<!--<img src="../../imgs/embedding.gif" style="width:300px;">-->

- Cargue los datos desde el archivo `"data/fashion2.csv"`
- Tome el  90%  del dataset como training y el restante para el test
- Realice la clasificación con: 
    - GaussianNB (Parametros por defecto)
    - Random Forest (Parametros por defecto)
    - SVM (kernel='rbf')
- La función debe devolver una lista con los modelos `[GN, RF, SV]` y una lista con los resultados `[score_GN, score_RF, score_SV]`


In [11]:
def clas_fashion():
    import numpy as np
    import pandas as pd
    from sklearn.model_selection import train_test_split
    from sklearn.naive_bayes import GaussianNB
    from sklearn.ensemble import RandomForestClassifier
    from sklearn.svm import SVC

    d = pd.read_csv("data/fashion2.csv",header=None)
    X = d.values[:,1:]
    y = d.values[:,0]
    X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=.10)
    GN=GaussianNB()
    RF=RandomForestClassifier()
    SV=SVC(kernel='rbf')
    GN.fit(X_train,y_train)
    RF.fit(X_train,y_train)  
    SV.fit(X_train,y_train) 
    score_GN = GN.score(X_test,y_test)
    score_RF = RF.score(X_test,y_test)
    score_SV = SV.score(X_test,y_test)
    
    return [GN, RF, SV], [score_GN, score_RF, score_SV]

#import urllib, inspect
#src8 = urllib.quote_plus(inspect.getsource(clas_fashion))
import urllib.parse, inspect
src8 = urllib.parse.quote_plus(inspect.getsource(clas_fashion))

In [12]:
import pandas as pd
FMnist = pd.read_csv("data/fashion2.csv",header=None)
print(FMnist.shape)
FMnist.head()

(2001, 785)


0    1    2    3    4    5    6    ...  778  779  780  781  782  783  784
0    0    1    2    3    4    5    6  ...  778  779  780  781  782  783  784
1    0    0    0    0    0    0    0  ...    0    0    0    0    0    0    0
2    1    0    0    0    0    0    0  ...    0    0    0    0    0    0    0
3    2    0    0    0    0    0    0  ...    0   63   53   31    0    0    0
4    2    0    0    0    0    0    0  ...    0  133  224  222   56    0    0

[5 rows x 785 columns]

In [13]:
models, scores = clas_fashion()
print("... comparison results ...")
print("GN: ", scores[0])
print("RF: ", scores[1])
print("SV: ", scores[2])

... comparison results ...
GN:  0.5024875621890548
RF:  0.8109452736318408
SV:  0.8009950248756219


In [14]:
!./run CHECK_SOLUTION PQ2_13 $src8

2020-08-01 16:22:02.148419: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
evaluation result CORRECT


---

### Ejercicio 4  <a name="eje4"></a>

Deep Learning con dataset **Fashion-MNIST**.

<!--<img src="../../imgs/embedding.gif" style="width:300px;">-->

- Cargue los datos desde el archivo `"data/fashion2.csv"`
- Tome el  90%  del dataset como training y el restante para el test
- Realice la clasificación con: 
    - DNN con la siguiente arquitectura
        - capa de entrada que reciba el vector de 784
        - Una capa oculta densa de 512 neuronas y activación ReLu
        - Una capa oculta densa de 128 neuronas y activación ReLu
        - Una capa de salida, según las clases del dataset. 
        - Entrene el modelo con la configuración: `epochs=5, batch_size=1000`
- La función debe devolver el modelo `DNN` y el resultado `score_DNN`

In [15]:
def clas_fashion2():
    import tensorflow as tf
    from tensorflow import keras
    import numpy as np
    import pandas as pd
    from sklearn.model_selection import train_test_split

    d = pd.read_csv("data/fashion2.csv",header=None)
    X = d.values[:,1:]
    y = d.values[:,0]
    X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=.10)
    model = keras.models.Sequential([
    keras.layers.Flatten( input_shape=(X_train[0].shape)),
    keras.layers.Dense(512, activation=tf.nn.relu),
    keras.layers.Dense(128, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax)
    ])

    model.compile(optimizer=tf.optimizers.Adam(),
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
    
    model.fit(X_train, y_train, epochs=5, batch_size=1000, verbose=0)
    test_loss, test_acc = model.evaluate(X_test, y_test)
   
    return model, test_acc

#import urllib, inspect
#src3 = urllib.quote_plus(inspect.getsource(clas_fashion2))
import urllib.parse, inspect
src3 = urllib.parse.quote_plus(inspect.getsource(clas_fashion2))

In [16]:
import pandas as pd
FMnist = pd.read_csv("data/fashion2.csv",header=None)
print(FMnist.shape)
FMnist.head()

(2001, 785)


0    1    2    3    4    5    6    ...  778  779  780  781  782  783  784
0    0    1    2    3    4    5    6  ...  778  779  780  781  782  783  784
1    0    0    0    0    0    0    0  ...    0    0    0    0    0    0    0
2    1    0    0    0    0    0    0  ...    0    0    0    0    0    0    0
3    2    0    0    0    0    0    0  ...    0   63   53   31    0    0    0
4    2    0    0    0    0    0    0  ...    0  133  224  222   56    0    0

[5 rows x 785 columns]

In [17]:
!./run CHECK_SOLUTION PQ2_14 $src3

2020-08-01 16:22:15.158656: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
2020-08-01 16:22:17.492773: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1
2020-08-01 16:22:17.496746: E tensorflow/stream_executor/cuda/cuda_driver.cc:313] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2020-08-01 16:22:17.496857: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (b07b89e2b908): /proc/driver/nvidia/version does not exist
2020-08-01 16:22:17.502329: I tensorflow/core/platform/profile_utils/cpu_utils.cc:102] CPU Frequency: 2300000000 Hz
2020-08-01 16:22:17.502596: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x1fdd100 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-08-01 16:22:17.502635: I tensorflow/com

---

#### <span style="color:#FFF; background-color:#080; font-weight:bold; padding:4px;">Importante</span> Pasos Finales.

#### 1. Guarde los cambios del taller:

In [18]:
%%javascript
require(["base/js/namespace"],function(Jupyter) {
    Jupyter.notebook.save_checkpoint();
});

<IPython.core.display.Javascript object>

#### 2. Reemplace la variable con su código de estudiante

In [20]:
import urllib.parse, inspect
taller = urllib.parse.quote_plus("quiz_02")
codigo = urllib.parse.quote_plus("q02_2170688")
print(taller, codigo)

!./convert_ipynb.sh $codigo $taller

quiz_02 q02_2170688
uno: q02_2170688 dos quiz_02
¡PERFECTO!, el archivo   'q02_2170688.ipynb'   se creó correctamente :D


#### 3. Envíe el archivo ""q02_CODIGOESTUDIANTE".ipynb" que se generó

---
# <center> Thanks! </center>